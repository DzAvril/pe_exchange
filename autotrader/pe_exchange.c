#include "pe_exchange.h"

#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <time.h>

#include "pe_common.h"
// Global variables
Product products[MAX_PRODUCTS];
int num_products = 0;
Trader traders[MAX_TRADERS];
int num_traders = MAX_TRADERS;

void load_products(const char* filename) {
  FILE* file = fopen(filename, "r");
  if (file == NULL) {
    perror("Error opening file");
    exit(EXIT_FAILURE);
  }
  fscanf(file, "%d", &num_products);
  for (int i = 0; i < num_products; i++) {
    fscanf(file, "%s", products[i].name);
  }
  fclose(file);
}

void parse_args(int argc, char** argv) {
  // printf("[debug] %s-%d Parsing args...\n", __FILE__, __LINE__);
  const char* filename = argv[1];
  // printf("[debug] %s-%d filename : %s.\n", __FILE__, __LINE__, filename);
  load_products(filename);
  num_traders = argc - 2;
}

void fork_child_process() {
  for (int i = 0; i < num_traders; i++) {
    sprintf(traders[i].exchange_fifo, FIFO_PATH_EXCHANGE_PREFIX, i);
    sprintf(traders[i].trader_fifo, FIFO_PATH_TRADER_PREFIX, i);
    // check if fifo already exists, remove it if so
    if (access(traders[i].exchange_fifo, F_OK) != -1) {
      if (remove(traders[i].exchange_fifo) != 0) {
        perror("Error removing FIFO");
        exit(EXIT_FAILURE);
      }
    }
    if (access(traders[i].trader_fifo, F_OK) != -1) {
      if (remove(traders[i].trader_fifo) != 0) {
        perror("Error removing FIFO");
        exit(EXIT_FAILURE);
      }
    }

    // create fifo
    if (mkfifo(traders[i].exchange_fifo, 0666) < 0) {
      perror("Error creating FIFO");
      exit(EXIT_FAILURE);
    }

    if (mkfifo(traders[i].trader_fifo, 0666) < 0) {
      perror("Error creating FIFO");
      exit(EXIT_FAILURE);
    }

    // fork a child process to run pe_trader
    pid_t pid = fork();
    if (pid == 0) {
      // child process
      char trader_id[10];
      sprintf(trader_id, "%d", i);
      char* trader_argv[] = {"./pe_trader", trader_id, NULL};
      printf("[PEX-Milestone] Launching trader pe_trader\n");
      execv(trader_argv[0], trader_argv);
    } else if (pid > 0) {
      // parent process
      traders[i].trader_id = i;
      traders[i].pid = pid;
    } else {
      perror("Error forking child process");
      exit(EXIT_FAILURE);
    }
  }
}

void teardown() {
  printf("[PEX-Milestone] Trader disconnected\n");
  for (int i = 0; i < num_traders; i++) {
    close(traders[i].exchange_fd);
    close(traders[i].trader_fd);
    unlink(traders[i].exchange_fifo);
    unlink(traders[i].trader_fifo);
    remove(traders[i].trader_fifo);
    remove(traders[i].exchange_fifo);
  }
}

void sig_handler(int sig, siginfo_t *info, void *context) {
  if (sig == SIGUSR1) {
    // char buf[MAX_MESSAGE_LENGTH];
    // int trader_id = -1;
    // for (int i = 0; i < num_traders; i++) {
    //   if (traders[i].pid == info->si_pid) {
    //     trader_id = i;
    //     break;
    //   }
    // }
    // ssize_t len = read(traders[trader_id].trader_fd, buf, sizeof(buf));
    // printf("[debug] %s:%d Received message: %.*s\n\n", __FILE__, __LINE__, (int)len, buf);
  } else {
    teardown();
    raise(sig);
  }
}

void connect_to_pipes() {
  printf("[PEX-Milestone] Opened Name Pipes\n");
  for (int i = 0; i < num_traders; i++) {
    traders[i].exchange_fd = open(traders[i].exchange_fifo, O_WRONLY);
    if (traders[i].exchange_fd == -1) {
      perror("Failed to open FIFO");
      exit(EXIT_FAILURE);
    }
    traders[i].trader_fd = open(traders[i].trader_fifo, O_RDONLY);
    if (traders[i].trader_fd == -1) {
      perror("Failed to open FIFO");
      exit(EXIT_FAILURE);
    }
  }
}

void send_message(int trader_id, const char* message) {
  printf("[PEX-Milestone] Exchange -> Trader: %s\n", message);
  size_t message_len = strlen(message);
  if (write(traders[trader_id].exchange_fd, message, message_len) != message_len) {
    perror("Error writing message to trader");
    exit(EXIT_FAILURE);
  }

  // send signal to trader
  if (kill(traders[trader_id].pid, SIGUSR1) == -1) {
    perror("Error sending signal to trader");
    exit(EXIT_FAILURE);
  }
}

void notify_market_open() {
  char buf[MAX_MESSAGE_LENGTH];
  memset(buf, '\0', sizeof(buf));
  sprintf(buf, MESSAGE_MARKET_OPEN);
  for (int i = 0; i < num_traders; i++) {
    send_message(i, buf);
  }
}

void serialize_response(Response* response, char* buf) {
  char type_string[MAX_TYPE_STRING_LENGTH];
  switch (response->type) {
    case BUY:
      strncpy(type_string, ORDER_TYPE_BUY, MAX_TYPE_STRING_LENGTH);
      break;
    case SELL:
      strncpy(type_string, ORDER_TYPE_SELL, MAX_TYPE_STRING_LENGTH);
      break;
    case AMEND:
      strncpy(type_string, ORDER_TYPE_AMEND, MAX_TYPE_STRING_LENGTH);
      break;
    case CANCEL:
      strncpy(type_string, ORDER_TYPE_CANCEL, MAX_TYPE_STRING_LENGTH);
      break;
    default:
      break;
  }
  sprintf(buf, "%s %s %s %d %d;", RESPONSE_PREFIX, type_string, response->product.name,
          response->quantity, response->price);
}

void nofify_market_sell() {
  char buf[MAX_MESSAGE_LENGTH];
  memset(buf, '\0', sizeof(buf));
  Response* res = (Response*)malloc(sizeof(Response));
  if (res == NULL) {
    perror("Failed to allocate memory.\n");
  }
  // randon create response
  srand(time(NULL));
  int random_product = rand() % num_products;
  res->type = SELL;
  res->product = products[random_product];
  res->quantity = rand() % MAX_ORDER_QUANTITY + 1;
  res->price = rand() % 1000;
  serialize_response(res, buf);
  // printf("[debug] %s-%d Send response: %s.\n", __FILE__, __LINE__, buf);
  for (int i = 0; i < num_traders; i++) {
    send_message(i, buf);
  }
  // release reponse
  free(res);
}

int main(int argc, char** argv) {
  // printf("Hello World from pe exchange!\n");
  if (argc > 1) {
    parse_args(argc, argv);
  }
  // set signal handler
  // signal(SIGUSR1, sig_handler);
  struct sigaction sa;
  memset(&sa, 0, sizeof(struct sigaction));
  sa.sa_sigaction = sig_handler;
  sa.sa_flags = SA_SIGINFO;
  sigemptyset(&sa.sa_mask);

  if (sigaction(SIGUSR1, &sa, NULL) == -1) {
      perror("sigaction");
      return 1;
  }

  // fork child processes for traders
  fork_child_process();

  // check if all traders are connected
  connect_to_pipes();

  // set MARKET OPEN to auto traders
  notify_market_open();

  // send MARKET SELL to auto traders
  nofify_market_sell();

  // wait for all child processes to exit
  int status;
  pid_t wpid;
  while ((wpid = wait(&status)) > 0) {
    if (WIFEXITED(status)) {
      printf("Child process %d terminated with exit status %d\n", wpid, WEXITSTATUS(status));
    } else if (WIFSIGNALED(status)) {
      printf("Child process %d terminated due to unhandled signal %d\n", wpid, WTERMSIG(status));
    }
  }
  teardown();
  return 0;
}