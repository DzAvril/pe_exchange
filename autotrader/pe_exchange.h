#ifndef PE_COMMON_H_
#define PE_COMMON_H_

#endif  // PE_COMMON_H_