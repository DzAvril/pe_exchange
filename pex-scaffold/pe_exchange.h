#ifndef PE_EXCHANGE_H
#define PE_EXCHANGE_H

#include "pe_common.h"

typedef struct {
  int index;
  int price;
  int quantity;
  int to_be_removed;
  OrderType type;
} OrderInfo;

typedef struct {
  OrderType type;
  int quantity;
  int price;
  int num_order;
} OrderBrief;

typedef struct {
  Product product;
  int num_product;
  int buy_level;
  int sell_level;
  OrderBrief orderBrief[MAX_ORDERS];
} Report;
#endif
