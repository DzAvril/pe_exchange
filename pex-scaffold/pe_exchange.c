#include "pe_exchange.h"

#include <sys/wait.h>
#include <time.h>

// Global variables
Product products[MAX_PRODUCTS];
int num_products = 0;
Trader traders[MAX_TRADERS];
int num_traders = MAX_TRADERS;
char *trader_path[MAX_TRADERS];
int exchange_fee_collected = 0;

void load_products(const char* filename) {
  FILE* file = fopen(filename, "r");
  if (file == NULL) {
    perror("Error opening file");
    exit(EXIT_FAILURE);
  }
  fscanf(file, "%d", &num_products);
  for (int i = 0; i < num_products; i++) {
    fscanf(file, "%s", products[i].name);
  }
  fclose(file);
  char all_products[PRODUCT_NAME_LENGTH * MAX_PRODUCTS];
  for (int i = 0; i < num_products; i++) {
    strcat(all_products, products[i].name);
    strcat(all_products, " ");
  }
  printf("%s Trading %d products: %s\n", LOG_EXCHANGE_PREFIX, num_products, all_products);
}

void parse_args(int argc, char** argv) {
  // printf("[debug] %s-%d Parsing args...\n", __FILE__, __LINE__);
  const char* filename = argv[1];
  // printf("[debug] %s-%d filename : %s.\n", __FILE__, __LINE__, filename);
  load_products(filename);
  num_traders = argc - 2;
  for (int i = 0; i < num_traders; i++) {
    trader_path[i] = argv[i + 2];
  }
}

void fork_child_process() {
  for (int i = 0; i < num_traders; i++) {
    sprintf(traders[i].exchange_fifo, FIFO_EXCHANGE, i);
    sprintf(traders[i].trader_fifo, FIFO_TRADER, i);
    // check if fifo already exists, remove it if so
    if (access(traders[i].exchange_fifo, F_OK) != -1) {
      if (remove(traders[i].exchange_fifo) != 0) {
        perror("Error removing FIFO");
        exit(EXIT_FAILURE);
      }
    }
    if (access(traders[i].trader_fifo, F_OK) != -1) {
      if (remove(traders[i].trader_fifo) != 0) {
        perror("Error removing FIFO");
        exit(EXIT_FAILURE);
      }
    }

    // create fifo
    if (mkfifo(traders[i].exchange_fifo, 0666) < 0) {
      perror("Error creating FIFO");
      exit(EXIT_FAILURE);
    }
    printf("%s Created FIFO %s\n", LOG_EXCHANGE_PREFIX, traders[i].exchange_fifo);

    if (mkfifo(traders[i].trader_fifo, 0666) < 0) {
      perror("Error creating FIFO");
      exit(EXIT_FAILURE);
    }
    printf("%s Created FIFO %s\n", LOG_EXCHANGE_PREFIX, traders[i].trader_fifo);

    // fork a child process to run pe_trader
    pid_t pid = fork();
    if (pid == 0) {
      // child process
      char trader_id[20];
      sprintf(trader_id, "%d", i);
      printf("%s Starting trader %d (%s)\n", LOG_EXCHANGE_PREFIX, i, trader_path[i]);
      char* trader_argv[] = {trader_path[i], trader_id, NULL};
      execv(trader_argv[0], trader_argv);
    } else if (pid > 0) {
      // parent process
      traders[i].trader_id = i;
      traders[i].pid = pid;
    } else {
      perror("Error forking child process");
      exit(EXIT_FAILURE);
    }
  }
}

void teardown() {
  for (int i = 0; i < num_traders; i++) {
    close(traders[i].exchange_fd);
    close(traders[i].trader_fd);
    unlink(traders[i].exchange_fifo);
    unlink(traders[i].trader_fifo);
    remove(traders[i].trader_fifo);
    remove(traders[i].exchange_fifo);
  }
}

void sig_handler(int sig, siginfo_t *info, void *context) {
  if (sig == SIGUSR1) {
    char buf[MAX_MESSAGE_LENGTH];
    int trader_id = -1;
    for (int i = 0; i < num_traders; i++) {
      if (traders[i].pid == info->si_pid) {
        trader_id = i;
        break;
      }
    }
    ssize_t len = read(traders[trader_id].trader_fd, buf, sizeof(buf));
    printf("[debug] %s:%d Received message: %.*s\n\n", __FILE__, __LINE__, (int)len, buf);
  } else {
    teardown();
    raise(sig);
  }
}

void connect_to_pipes() {
  for (int i = 0; i < num_traders; i++) {
    traders[i].exchange_fd = open(traders[i].exchange_fifo, O_WRONLY);
    if (traders[i].exchange_fd == -1) {
      perror("Failed to open FIFO");
      exit(EXIT_FAILURE);
    }
    traders[i].trader_fd = open(traders[i].trader_fifo, O_RDONLY);
    if (traders[i].trader_fd == -1) {
      perror("Failed to open FIFO");
      exit(EXIT_FAILURE);
    }
  }
}

void send_message(int trader_id, const char* message) {
  size_t message_len = strlen(message);
  if (write(traders[trader_id].exchange_fd, message, message_len) != message_len) {
    perror("Error writing message to trader");
    exit(EXIT_FAILURE);
  }

  // send signal to trader
  if (kill(traders[trader_id].pid, SIGUSR1) == -1) {
    perror("Error sending signal to trader");
    exit(EXIT_FAILURE);
  }
}

void notify_market_open() {
  char buf[MAX_MESSAGE_LENGTH];
  memset(buf, '\0', sizeof(buf));
  sprintf(buf, MESSAGE_MARKET_OPEN);
  for (int i = 0; i < num_traders; i++) {
    send_message(i, buf);
  }
}

void serialize_response(Response* response, char* buf) {
  char type_string[MAX_TYPE_STRING_LENGTH];
  switch (response->type) {
    case BUY:
      strncpy(type_string, ORDER_TYPE_BUY, MAX_TYPE_STRING_LENGTH);
      break;
    case SELL:
      strncpy(type_string, ORDER_TYPE_SELL, MAX_TYPE_STRING_LENGTH);
      break;
    case AMEND:
      strncpy(type_string, ORDER_TYPE_AMEND, MAX_TYPE_STRING_LENGTH);
      break;
    case CANCEL:
      strncpy(type_string, ORDER_TYPE_CANCEL, MAX_TYPE_STRING_LENGTH);
      break;
    default:
      break;
  }
  sprintf(buf, "%s %s %s %d %d;", RESPONSE_PREFIX, type_string, response->product.name,
          response->quantity, response->price);
}

// void nofify_market_sell() {
//   char buf[MAX_MESSAGE_LENGTH];
//   memset(buf, '\0', sizeof(buf));
//   Response* res = (Response*)malloc(sizeof(Response));
//   if (res == NULL) {
//     perror("Failed to allocate memory.\n");
//   }
//   // randon create response
//   srand(time(NULL));
//   int random_product = rand() % num_products;
//   res->type = SELL;
//   res->product = products[random_product];
//   res->quantity = rand() % MAX_ORDER_QUANTITY + 1;
//   res->price = rand() % 1000;
//   serialize_response(res, buf);
//   // printf("[debug] %s-%d Send response: %s.\n", __FILE__, __LINE__, buf);
//   for (int i = 0; i < num_traders; i++) {
//     send_message(i, buf);
//   }
//   // release reponse
//   free(res);
// }

int main(int argc, char** argv) {
  printf("%s Starting\n", LOG_EXCHANGE_PREFIX);
  if (argc > 1) {
    parse_args(argc, argv);
  }
  // set signal handler
  struct sigaction sa;
  memset(&sa, 0, sizeof(struct sigaction));
  sa.sa_sigaction = sig_handler;
  sa.sa_flags = SA_SIGINFO;
  sigemptyset(&sa.sa_mask);

  if (sigaction(SIGUSR1, &sa, NULL) == -1) {
      perror("sigaction");
      return 1;
  }


  // fork child processes for traders
  fork_child_process();

  // check if all traders are connected
  connect_to_pipes();

  // set MARKET OPEN to auto traders
  notify_market_open();

  // send MARKET SELL to auto traders
  // nofify_market_sell();

  // wait for all child processes to exit
  int status;
  pid_t wpid;
  while ((wpid = wait(&status)) > 0) {
    // if (WIFEXITED(status)) {
    //   printf("Child process %d terminated with exit status %d\n", wpid, WEXITSTATUS(status));
    // } else if (WIFSIGNALED(status)) {
    //   printf("Child process %d terminated due to unhandled signal %d\n", wpid, WTERMSIG(status));
    // }
    int trader_id = -1;
    for (int i = 0; i < num_traders; i++) {
      if (traders[i].pid == wpid) {
        trader_id = traders[i].trader_id;
        break;
      }
    }
    printf("%s Trader %d disconnected\n", LOG_EXCHANGE_PREFIX, trader_id);
  }
  printf("%s Trading completed\n", LOG_EXCHANGE_PREFIX);
  printf("%s Exchange fees collected: $%d\n", LOG_EXCHANGE_PREFIX, exchange_fee_collected);
  teardown();
  return 0;
}