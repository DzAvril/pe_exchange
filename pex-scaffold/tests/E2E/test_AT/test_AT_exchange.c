#define _POSIX_C_SOURCE 199309L

#include "../../../pe_exchange.h"

// Global variables
extern Product products[MAX_PRODUCTS];
extern int num_products;
extern Trader traders[MAX_TRADERS];
extern int num_traders;
extern char* trader_path[MAX_TRADERS];
extern int exchange_fee_collected;
extern Order orderbook[MAX_ORDERS];
extern int num_orders;

void sig_handler(int sig, siginfo_t* info, void* context) {
  if (sig == SIGUSR1) {
    char buf[MAX_MESSAGE_LENGTH];
    int trader_index = get_trader_by_pid(info->si_pid);
    if (trader_index == -1) {
      perror("Error getting trader by pid");
      exit(EXIT_FAILURE);
    }
    read(traders[trader_index].trader_fd, buf, sizeof(buf));
    printf("%s Received message from trader: %s.\n", LOG_EXCHANGE_PREFIX, buf);
  }
}

void nofify_market_sell() {
  char buf[MAX_MESSAGE_LENGTH];
  memset(buf, '\0', sizeof(buf));
  Response* res = (Response*)malloc(sizeof(Response));
  if (res == NULL) {
    perror("Failed to allocate memory.\n");
  }
  // randon create response
  srand(time(NULL));
  int random_product = rand() % num_products;
  res->type = SELL;
  res->product = products[random_product];
  res->quantity = rand() % MAX_ORDER_QUANTITY + 1;
  res->price = rand() % 1000;
  serialize_response(res, buf);
  // printf("[debug] %s-%d Send response: %s.\n", __FILE__, __LINE__, buf);
  for (int i = 0; i < num_traders; i++) {
    send_message(i, buf);
  }
  // release reponse
  free(res);
}

int main(int argc, char** argv) {
  printf("%s Starting\n", LOG_EXCHANGE_PREFIX);
  if (argc > 1) {
    parse_args(argc, argv);
  }
  // set signal handler
  struct sigaction sa;
  memset(&sa, 0, sizeof(struct sigaction));
  sa.sa_sigaction = sig_handler;
  sa.sa_flags = SA_SIGINFO;
  sigemptyset(&sa.sa_mask);

  if (sigaction(SIGUSR1, &sa, NULL) == -1) {
    perror("sigaction");
    return 1;
  }

  // if (sigaction(SIGCHLD, &sa, NULL) == -1) {
  //   perror("sigaction");
  //   return 1;
  // }

  // fork child processes for traders
  fork_child_process();

  // check if all traders are connected
  // connect_to_pipes();

  // set MARKET OPEN to auto traders
  notify_market_open();

  // send MARKET SELL to auto traders
  nofify_market_sell();

  int epollfd = epoll_create1(0);
  if (epollfd == -1) {
    perror("epoll_create failed");
    exit(EXIT_FAILURE);
  }
  struct epoll_event ev[MAX_TRADERS];
  for (int i = 0; i < num_traders; i++) {
    ev[i].events = EPOLLIN | EPOLLHUP | EPOLLET;
    ev[i].data.fd = traders[i].trader_fd;
    if (epoll_ctl(epollfd, EPOLL_CTL_ADD, traders[i].trader_fd, &ev[i]) == -1) {
      perror("epoll_ctl failed");
      exit(EXIT_FAILURE);
    }
  }
  int all_fifo_closed = 0;
  // listen event of traders close fifo
  while (1) {
    int nfds;
    if ((nfds = epoll_wait(epollfd, ev, num_traders, -1)) != -1) {
      for (int i = 0; i < nfds; i++) {
        if (ev[i].events & EPOLLHUP) {
          int trader_index = get_trader_by_fd(ev[i].data.fd);
          if (trader_index == -1) {
            perror("Error getting trader by fd");
            exit(EXIT_FAILURE);
          }
          printf("%s Trader %d disconnected\n", LOG_EXCHANGE_PREFIX,
                 traders[trader_index].trader_id);

          cleanup_trader(trader_index);
          all_fifo_closed = 1;
          for (int j = 0; j < num_traders; j++) {
            if (open(traders[j].trader_fifo, O_RDONLY) != -1) {
              all_fifo_closed = 0;
              break;
            }
          }
        }
      }
    }
    int status;
    pid_t wpid;
    int all_children_exited = 1;
    while ((wpid = waitpid(-1, &status, WNOHANG)) > 0) {
      if (!WIFEXITED(status)) {
        all_children_exited = 0;
      }
    }
    if (all_children_exited && all_fifo_closed) {
      break;
    }
  }
  printf("%s Trading completed\n", LOG_EXCHANGE_PREFIX);
  printf("%s Exchange fees collected: $%d\n", LOG_EXCHANGE_PREFIX, exchange_fee_collected);
  teardown();
  return 0;
}