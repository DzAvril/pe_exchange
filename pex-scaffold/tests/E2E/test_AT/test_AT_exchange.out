[PEX] Received message from exchange: MARKET OPEN;.
[PEX] Received message from exchange: MARKET SELL GPU 356 318;.
[PEX] Starting
[PEX] Trading 4 products: GPU CPU Memory switch
[PEX] Created FIFO /tmp/pe_exchange_0
[PEX] Created FIFO /tmp/pe_trader_0
[PEX] Connected to /tmp/pe_exchange_0
[PEX] Connected to /tmp/pe_trader_0
[PEX] Received message from trader: BUY 0 GPU 356 318;.
[PEX] Trader 0 disconnected
[PEX] Trading completed
[PEX] Exchange fees collected: $0
